New ToolTipClosing, ToolTipOpening attached events and ToolTip.Opened, ToolTip.Closed

maxkatz6 opened the pull request:
## What does the pull request do?

```diff
public class ToolTip
{
+     public static readonly RoutedEvent ToolTipOpeningEvent;
+     public static readonly RoutedEvent ToolTipClosingEvent;
+     public static void AddToolTipOpeningHandler(Control element, EventHandler<RoutedEventArgs> handler);
+     public static void RemoveToolTipOpeningHandler(Control element, EventHandler<RoutedEventArgs> handler);
+     public static void AddToolTipClosingHandler(Control element, EventHandler<RoutedEventArgs> handler);
+     public static void RemoveToolTipClosingHandler(Control element, EventHandler<RoutedEventArgs> handler);
}
```

Behavior is ported from WPF, but implementation is different, just like our ToolTip API is already quite different. Most of this decisions were made for consistency with the rest of the framework.

Common:
1. When ToolTipOpening is raised, it's possible to replace TipProperty, which will cause opening of another tip.
2. It's also possible to set ToolTipOpening Handled=true, which will essentially cancel tooltip opening.
3. It's not possible to cancel ToolTipClosing in any way. 

Different:
1. Instead of adding attached events to the ToolTipService, I added them to ToolTip class. Following consistency with other attached ToolTip APIs.
2. Same about adding new members to the Control class - we never added anything ToolTip (or ContextMenu) related to the Control type itself - everything is attached.
3. Opened/Closed events are direct-routed in WPF, but CLR events in this PR. Following the same behavior with ContextMenu and Flyout events.

To raise Opening/Closing events, I used IsOpen.Coerse callback, which is also handy to cancel tooltip opening. In WPF these events were only usable via ToolTipService, and handled there. In general, in Avalonia tooltip is completely usable without ToolTipService, so these events should as well.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Fixed issues

Fixes #15232

maxkatz6 committed:
Add ToolTip.Opened and ToolTip.Closed

maxkatz6 committed:
Add ToolTipOpeningEvent and ToolTipClosingEvent attache events

maxkatz6 committed:
Add tests and small sample

maxkatz6 committed:
Docs

BAndysc commented:
In case of custom drawn controls, tooltip content may depend on the position of the cursor thus it would be very useful to have a location of the pointer in ToolTipOpeningEvent that is causing the tooltip to show up if it is possible

MrJul commented:
> 3\. It's not possible to cancel ToolTipClosing in any way.

Is it a technical limitation? I can see uses in preventing tooltips from closing.

maxkatz6 commented:
@MrJul I am just following WPF behavior here. Technically, it can be done. 

@BAndysc would be a bit tricky to pass position from ToolTipService to ToolTip.

MrJul commented:
Should we unify the API between all popup-like controls here?

Having all necessary events (`Opening`, `Opened`, `Closing`, `Closed`) on `Popup` and re-trigger them on `ContextMenu`, `ToolTip` and `Flyout`. Currently it feels a bit random, and the signature sometimes differ between controls (e.g. `CancelEventArgs` vs standard event args).

MrJul commented on src/Avalonia.Controls/ToolTip.cs:
```diff
@@ -297,26 +349,76 @@ private static void IsOpenChanged(AvaloniaPropertyChangedEventArgs e)
+                if (isOpen)
+                {
+                    var args = new RoutedEventArgs(ToolTipOpeningEvent);
+                    control.RaiseEvent(args);
```
I'll let @grokys comment on this, but I'm a bit worried of having such side effects in coercion methods, potentially leading to reentrancy or other hard to debug problems. Coercion methods are called in the middle of setting an `EffectiveValue`, and I don't think this code currently handle reentrancy. (For example, `PropertyChanged` isn't called until all the internal effective value bookkeeping is done to avoid issues.)

maxkatz6 replied:
Yeah, didn't think about reentrancy. It's technically possible to change IsOpen value inside.

TomEdwardsEnscape replied:
It's also possible for the events to be raised when the value isn't changing. Aside from any internal coercion calls, the user can do this at any time:

```
myControl.CoerceValue(ToolTip.IsOpenProperty);
```

`ToolTip.IsOpenChanged` is the right place for this code. It's a bit weird that `IsOpen` will be true when `ToolTipOpening` is raised, but the event name makes it very clear what's going on.

robloo commented:
> Should we unify the API between all popup-like controls here?
> 
> Having all necessary events (`Opening`, `Opened`, `Closing`, `Closed`) on `Popup` and re-trigger them on `ContextMenu`, `ToolTip` and `Flyout`. Currently it feels a bit random, and the signature sometimes differ between controls (e.g. `CancelEventArgs` vs standard event args).

I'm a big fan of unifying APIs everywhere it's possible. This is a great place to do it IMO.

Also note that `Expander` is related to this discussion and while the naming is different (Expanding/Collapsing/etc.) usage and API patterns should be the same as the rest.

billhenn commented:
@maxkatz6 The update works well.  I was able to dynamically update a tooltip before it displayed.

I agree that having a consistent API for the `Opening`, `Opened`, `Closing`, and `Closed` events in various places would be ideal.

I'm also usually a fan of using something like `e.Cancel` to prevent a default behavior.  `e.Handled` is technically there to mean stop routing an event further.  The way WPF uses it for `ToolTipOpening` events is sort of a gray area, since they hijack `e.Handled` to say whether to cancel the default functionality.  One could probably argue either way on that usage.

robloo commented:
@billhenn @maxkatz6 

> I'm also usually a fan of using something like `e.Cancel` to prevent a default behavior. `e.Handled` is technically there to mean stop routing an event further. The way WPF uses it for `ToolTipOpening` events is sort of a gray area, since they hijack `e.Handled` to say whether to cancel the default functionality. One could probably argue either way on that usage.

Note that this exact topic was discussed for Expander here: https://github.com/AvaloniaUI/Avalonia/pull/9979#discussion_r1070724369. The agreement was to add `CancelRoutedEventArgs` to go along with the existing `CancelRoutedEventArgs` establishing Avalonia's convention. We differ here from WPF but it's better for the same reasons you state. All controls should be using this convention -- tooltip as well.

Just keep in mind Avalonia has these cancel event args already and 3rd party libraries should ideally use them as well.

robloo commented on src/Avalonia.Controls/ToolTip.cs:
```diff
@@ -80,6 +81,28 @@ public class ToolTip : ContentControl, IPopupHostProvider
+        /// ToolTipOpening will not be raised if the value of ToolTip is null or otherwise unset. Do not deliberately set ToolTip to null while a tooltip is open or opening; this will not have the effect of closing the tooltip, and will instead create an undesirable visual artifact in the UI.
+        /// </remarks>
+        public static readonly RoutedEvent ToolTipOpeningEvent =
+            RoutedEvent.Register<ToolTip, RoutedEventArgs>("ToolTipOpening", RoutingStrategies.Direct);
```
As discussed new events here should follow existing conventions and use `CancelRoutedEventArgs`. This follows Expander which itself was based on what Flyout did (although flyout isn't API symmetric and doesn't allow cancelling an open).

jmacato commented:
@cla-avalonia recheck

cla-avalonia commented:
@maxkatz6,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

maxkatz6 committed:
Remove new Opened/Closed events

maxkatz6 committed:
Merge remote-tracking branch 'origin/master' into tooltip-opening-closing

# Conflicts:
#	tests/Avalonia.Controls.UnitTests/ToolTipTests.cs

maxkatz6 committed:
Update tests

maxkatz6 committed:
Restore internal Closed

maxkatz6 committed:
Update tests

maxkatz6 committed:
Don't use coerce logic, use CancelRoutedEventArgs in Opening API

maxkatz6 committed:
Update samples API

maxkatz6 commented:
Updated PR:
1. Reimplemented attached events to not rely on Coerce callback
2. Changed Opening args type from RoutedEventArgs to CancelRoutedEventArgs
3. Removed Opened and Closed events. These weren't part of the initial issue, and can be re-added in the future. It's likely involve breaking changes in ContextMenu or Flyout controls, as these have different events. Also, Popup itself wasn't really intended to be cancellable.

Note, Closing event is still not cancellable. If necessary, it can be implemented later.

billhenn commented:
@maxkatz6 Thanks, I tried the PR build and it does what we needed for the tooltip opening notification.

jmacato committed:
Merge branch 'master' into tooltip-opening-closing

maxkatz6 committed:
Merge branch 'master' into tooltip-opening-closing

maxkatz6 committed:
Merge branch 'master' into tooltip-opening-closing

maxkatz6 committed:
Fix incorrect routed event definition
